### PR TITLE
Fix building cuspatial with CCCL main

### PR DESCRIPTION
**Pull request by trxcllnt**
Similar to https://github.com/rapidsai/cudf/pull/15552, we are testing [building RAPIDS with CCCL's main branch](https://github.com/NVIDIA/cccl/pull/1667) to get ahead of any breaking changes.

### CODE REVIEW

**Comment by harrism** [comment]
I think you mean cuspatial? :)

File: cpp/include/cuspatial/detail/index/construction/phase_2.cuh [hunk 1]
```diff
@@ -85,9 +85,9 @@ inline rmm::device_uvector<uint32_t> flatten_point_keys(
                     keys_and_levels + num_valid_nodes,
                     flattened_keys.begin(),
                     [last_level = max_depth - 1] __device__(auto const& val) {
-                      bool is_parent{false};
-                      uint32_t key{}, level{};
-                      thrust::tie(key, level, is_parent) = val;
+                      auto& key       = thrust::get<0>(val);
+                      auto& level     = thrust::get<1>(val);
+                      auto& is_parent = thrust::get<2>(val);
```
Comment (harrism) [inline cpp/include/cuspatial/detail/index/construction/phase_2.cuh:90]: structured binding didn't work?

Comment (trxcllnt) [reply cpp/include/cuspatial/detail/index/construction/phase_2.cuh:90]: Structured binding works with new CCCL 2.5, but not current CCCL 2.2.

File: .github/workflows/pr.yaml [hunk 2]
```diff
@@ -110,6 +110,24 @@ jobs:
...
+        # Tell rapids-cmake to use custom CCCL and cuCollections forks
+        rapids_branch="$(yq '.x-git-defaults.tag' /opt/rapids-build-utils/manifest.yaml)";
+        rapids_version="${rapids_branch#branch-}";
+        curl -fsSL -o- https://raw.githubusercontent.com/trxcllnt/rapids-cmake/branch-24.04-cccl-2.4.0/rapids-cmake/cpm/patches/cccl/revert_pr_211.diff \
+          | tee ~/rapids-cmake-revert_pr_211.diff;
+        curl -fsSL -o- "https://raw.githubusercontent.com/rapidsai/rapids-cmake/${rapids_branch}/rapids-cmake/cpm/versions.json" \
+          | jq -r ".packages.CCCL *= {\"version\": \"2.5.0\", \"git_tag\": \"pull-request/1667\"}" \
+          | jq -r "(.packages.CCCL.patches[] | select(.file == \"cccl/revert_pr_211.diff\")).file = \"${HOME}/rapids-cmake-revert_pr_211.diff\"" \
+          | jq -r ".packages.cuco *= {\"git_url\": \"https://github.com/trxcllnt/cuCollections.git\", \"git_tag\": \"rapids-${rapids_version}-cccl-2.5.0\", \"always_download\": true}" \
+          | tee ~/rapids-cmake-override-versions.json;
```
Comment (trxcllnt) [inline .github/workflows/pr.yaml:122]: Need to remove all this before merging

```suggestion
```

File: .github/workflows/pr.yaml [hunk 3]
```diff
@@ -110,6 +110,24 @@ jobs:
...
-        build-all -DBUILD_TESTS=ON -DBUILD_BENCHMARKS=ON --verbose;
+        build-all \
+          -j$(nproc --ignore=1) -v \
+          -DBUILD_TESTS=ON \
+          -DBUILD_BENCHMARKS=ON \
+          -DCUSPATIAL_CUDF_GIT_REPO="https://github.com/trxcllnt/cudf.git" \
+          -DCUSPATIAL_CUDF_GIT_TAG="fix/cccl-2.5" \
+          -DCMAKE_CXX_FLAGS="-ftemplate-backtrace-limit=0" \
+          -DCMAKE_CUDA_FLAGS="-ftemplate-backtrace-limit=0" \
+          -DRAPIDS_CMAKE_CPM_DEFAULT_VERSION_FILE="${HOME}/rapids-cmake-override-versions.json";
```
Comment (trxcllnt) [inline .github/workflows/pr.yaml:132]: Need to revert this before merging

```suggestion
        build-all -DBUILD_TESTS=ON -DBUILD_BENCHMARKS=ON --verbose;
```

**Review by harrism — APPROVED** [review]

File: cpp/cmake/thirdparty/get_cudf.cmake [hunk 4]
```diff
@@ -62,11 +62,11 @@ endfunction()
 set(CUSPATIAL_MIN_VERSION_cudf "${CUSPATIAL_VERSION_MAJOR}.${CUSPATIAL_VERSION_MINOR}")
 
 if(NOT DEFINED CUSPATIAL_CUDF_GIT_REPO)
-  set(CUSPATIAL_CUDF_GIT_REPO https://github.com/rapidsai/cudf.git)
+  set(CUSPATIAL_CUDF_GIT_REPO https://github.com/trxcllnt/cudf.git)
```
Comment (trxcllnt) [inline cpp/cmake/thirdparty/get_cudf.cmake:65]: Need to revert this before merging:

```suggestion
  set(CUSPATIAL_CUDF_GIT_REPO https://github.com/rapidsai/cudf.git)
```

File: cpp/cmake/thirdparty/get_cudf.cmake [hunk 5]
```diff
@@ -62,11 +62,11 @@ endfunction()
...
 endif()
 
 if(NOT DEFINED CUSPATIAL_CUDF_GIT_TAG)
-  set(CUSPATIAL_CUDF_GIT_TAG branch-${CUSPATIAL_MIN_VERSION_cudf})
+  set(CUSPATIAL_CUDF_GIT_TAG fix/cccl-2.5)
```
Comment (trxcllnt) [inline cpp/cmake/thirdparty/get_cudf.cmake:69]: Need to revert this before merging:

```suggestion
  set(CUSPATIAL_CUDF_GIT_TAG branch-${CUSPATIAL_MIN_VERSION_cudf})
```

File: rapids_config.cmake [hunk 6]
```diff
@@ -26,6 +26,9 @@ else()
   )
 endif()
 
+set(rapids-cmake-repo trxcllnt/rapids-cmake)
+set(rapids-cmake-branch fea/cccl-2.5)
+
```
Comment (trxcllnt) [inline rapids_config.cmake:31]: Need to revert this before merging:

```suggestion
```

**Comment by harrism** [comment]
Looks like we have conda-notebook-tests failures on both of your current cuSpatial PRs?

**Pull request merged**